vignettes/invariants.Rmd

```diff
@@ -190,38 +187,50 @@ df2[[2]]
 .subset2(df2, 2)
 ```
 
-### Corollary
+NB: `x[[j]]` always returns an object of size `nrow(x)` if the column exists.
 
-Accessing `NA` or missing columns returns `NULL`.
-Out-of-bounds indexing throws an error.
+`j` must be a single number or a string, as enforced by `vec_as_index(j)`.
+
+```{r double-bracket-requires-scalar-j-index, dftbl = TRUE}
+df[[1:2]]
+df[[c("n", "c")]]
+df[[TRUE]]
+df[[mean]]
+```
+
+<!-- HW: the failure with mean needs a better error. -->
+
+Missing and out-of-bounds (OOB) values of `j` return `NULL` with a warning or error:
 
 ```{r double-bracket-j-na, dftbl = TRUE}
-df[[NA]]
-df[[NA_integer_]]
 df[[NA_character_]]
 df[["x"]]
+
+df[[NA_integer_]]
+df[[-1]]
 df[[4]]
 ```
 
-`[[` requires scalar `vec_as_index(j)`.
+<!-- HW: why is x[[NA]] an OOB error and not a type error? -->
 
-```{r double-bracket-requires-scalar-j-index, dftbl = TRUE}
-df[[1]]
-df[[1:2]]
-df[[1:3]]
-df[["n"]]
-df[[c("n", "c")]]
-df[[TRUE]]
-df[[FALSE]]
+OOB access only generate a warning because a common existing idioms is to check for the absense of a column with `is.null(df$x)`.
+
+Non-integer numeric values produce an error:
+
+```{r, double-bracket-numeric, dftbl = TRUE}
+df[[1.5]]
 ```
 
-`x[[j]]` always returns an object of size `nrow(x)` if the column exists.
+## Definition of `x$name`
+
+`x$name` is equal to `x[["name"]]`.
```

Comment:
I'm using "equivalent" to indicate that this holds both for subsetting and subassignment, "equal" seems weaker. Is this important?

We could keep one subset and one subassign example. I agree that the battery of examples is overkill here.

Reply:
I used equals to match all the wording in vctrs